include/exadg/convection_diffusion/driver.cpp

```diff
@@ -143,28 +144,124 @@ Driver<dim, Number>::ale_update() const
   time_int_bdf->ale_update();
 }
 
+template<int dim, typename Number>
+void
+Driver<dim, Number>::mark_cells_coarsening_and_refinement(dealii::Triangulation<dim> & tria,
+                                                          VectorType const & solution) const
+{
+  mark_cells_kelly_error_estimator(tria,
+                                   pde_operator->get_dof_handler(),
+                                   pde_operator->get_constraints(),
+                                   *pde_operator->get_mapping(),
+                                   solution,
+                                   application->get_parameters().degree +
+                                     1 /* n_face_quadrature_points */,
+                                   application->get_parameters().amr_data);
+}
+
+template<int dim, typename Number>
+void
+Driver<dim, Number>::setup_after_coarsening_and_refinement()
+{
+  // Update mapping
+  AssertThrow(ale_mapping.get() == 0,
+              dealii::ExcMessage(
+                "Combination of adaptive mesh refinement and ALE not implemented."));
+
+  std::shared_ptr<dealii::MappingQCache<dim>> mapping_q_cache =
+    std::dynamic_pointer_cast<dealii::MappingQCache<dim>>(mapping);
+  AssertThrow(
+    mapping_q_cache.get() == 0,
+    dealii::ExcMessage(
+      "Combination of adaptive mesh refinement and dealii::MappingQCache not implemented."));
+
+  pde_operator->setup_after_coarsening_and_refinement();
+
+  postprocessor->setup(*pde_operator);
+}
+
+
+template<int dim, typename Number>
+void
+Driver<dim, Number>::do_adaptive_refinement(unsigned int const time_step_number)
```

Comment:
I did not fully understand this comment: https://github.com/exadg/exadg/pull/586#discussion_r1386303815

Is the idea that this function can treat both the steady and unsteady case? If so, one would move the next if statement `if(trigger_coarsening_and_refinement_now(time_step_size))` outside this function (which has the advantage that we do not need to pass the `time_step_number` to this function), and introduce another if-else somewhat below (I will make another comment ...)

Reply:
Now, `Driver::do_adaptive_refinement` might be used for `ProblemType::Steady` and `ProblemType::Unsteady` problems. We lack the `driver_steady->prepare_coarsening_and_refinement()` and `driver_steady->interpolate_after_coarsening_after_refinement()`. I asserted the relevant lines in `Driver::do:adaptive_refinement`.